flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushPartitionIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,325 @@
+		Filter filter = call.rel(0);
+		LogicalTableScan scan = call.rel(1);
+		FlinkContext context = call.getPlanner().getContext().unwrap(FlinkContext.class);
+		TableSourceTable tableSourceTable = scan.getTable().unwrap(TableSourceTable.class);
+
+		// build pruner
+		RelDataType inputFieldTypes = filter.getInput().getRowType();
+		List<String> inputFieldNames = inputFieldTypes.getFieldNames();
+		List<String> partitionFieldNames = tableSourceTable.catalogTable().getPartitionKeys();
+		RelBuilder relBuilder = call.builder();
+		RexBuilder rexBuilder = relBuilder.getRexBuilder();
+		Tuple2<Seq<RexNode>, Seq<RexNode>> allPredicates = RexNodeExtractor.extractPartitionPredicateList(
+			filter.getCondition(),
+			FlinkRelOptUtil.getMaxCnfNodeCount(scan),
+			inputFieldNames.toArray(new String[0]),
+			rexBuilder,
+			partitionFieldNames.toArray(new String[0])
+			);
+		RexNode partitionPredicate = RexUtil.composeConjunction(rexBuilder, JavaConversions.seqAsJavaList(allPredicates._1));
+
+		if (partitionPredicate.isAlwaysTrue()){
+			return;
+		}
+
+		List<LogicalType> partitionFieldTypes = partitionFieldNames.stream().map(name -> {
+			int index  = inputFieldNames.indexOf(name);
+			if (index < 0) {
+				throw new TableException(String.format("Partitioned key '%s' isn't found in input columns. " +
+					"Validator should have checked that.", name));
+			}			return inputFieldTypes.getFieldList().get(index).getType(); })
+			.map(FlinkTypeFactory::toLogicalType).collect(Collectors.toList());
+
+		RexNode finalPartitionPredicate = adjustPartitionPredicate(inputFieldNames, partitionFieldNames, partitionPredicate);
+		Function<List<Map<String, String>>, List<Map<String, String>>> defaultPruner = partitions -> PartitionPruner.prunePartitions(
+			context.getTableConfig(),
+			partitionFieldNames.toArray(new String[0]),
+			partitionFieldTypes.toArray(new LogicalType[0]),
+			partitions,
+			finalPartitionPredicate);
+
+		// get partitions from table/catalog and prune
+		Optional<Catalog> catalogOptional = context.getCatalogManager().getCatalog(tableSourceTable.tableIdentifier().getCatalogName());
+		List<Map<String, String>> remainingPartitions = null;
+		Optional<List<Map<String, String>>> optionalPartitions;
+		// fields to read partitions from catalog and build new statistic
+		DynamicTableSource dynamicTableSource = tableSourceTable.tableSource().copy();
+		ObjectIdentifier identifier = tableSourceTable.tableIdentifier();
+		ObjectPath tablePath = identifier.toObjectPath();
+		try {
+			optionalPartitions = ((SupportsPartitionPushDown) dynamicTableSource).listPartitions();
+			if (optionalPartitions.isPresent() && !optionalPartitions.get().isEmpty()) {
+				remainingPartitions = defaultPruner.apply(optionalPartitions.get());
+			}
+		} catch (UnsupportedOperationException e) {
+			// check catalog whether is available
+			// we will read partitions from catalog if table doesn't support listPartitions.
+			if (!catalogOptional.isPresent()){
```

Comment:
nit: please add a blank between `)`and `{`, there are many similar case: line 240, line 250, etc